evalml/tests/pipeline_tests/test_pipelines.py

```diff
@@ -2323,3 +2324,93 @@ def test_predict_has_input_target_name(problem_type, X_y_binary, X_y_multi, X_y_
     else:
         predictions = clf.predict(X)
     assert predictions.name == "test target name"
+
+
+def test_linear_pipeline_iteration(logistic_regression_binary_pipeline_class):
+    expected_order = [Imputer(), OneHotEncoder(), StandardScaler(), LogisticRegressionClassifier()]
+
+    pipeline = logistic_regression_binary_pipeline_class({})
+    order = [c for c in pipeline]
+    order_again = [c for c in pipeline]
+
+    assert order == expected_order
+    assert order_again == expected_order
+
+    expected_order_params = [Imputer(numeric_impute_strategy='median'), OneHotEncoder(top_n=2), StandardScaler(), LogisticRegressionClassifier()]
+
+    pipeline = logistic_regression_binary_pipeline_class({'One Hot Encoder': {'top_n': 2}, 'Imputer': {'numeric_impute_strategy': 'median'}})
+    order_params = [c for c in pipeline]
+    order_again_params = [c for c in pipeline]
+
+    assert order_params == expected_order_params
+    assert order_again_params == expected_order_params
+
+
+def test_nonlinear_pipeline_iteration(nonlinear_binary_pipeline_class):
+    expected_order = [Imputer(), OneHotEncoder(), ElasticNetClassifier(), OneHotEncoder(), RandomForestClassifier(), LogisticRegressionClassifier()]
+
+    pipeline = nonlinear_binary_pipeline_class({})
+    order = [c for c in pipeline]
+    order_again = [c for c in pipeline]
+
+    assert order == expected_order
+    assert order_again == expected_order
+
+    expected_order_params = [Imputer(), OneHotEncoder(top_n=2), ElasticNetClassifier(), OneHotEncoder(top_n=5), RandomForestClassifier(), LogisticRegressionClassifier()]
+
+    pipeline = nonlinear_binary_pipeline_class({'OneHot_ElasticNet': {'top_n': 2}, 'OneHot_RandomForest': {'top_n': 5}})
+    order_params = [c for c in pipeline]
+    order_again_params = [c for c in pipeline]
+
+    assert order_params == expected_order_params
+    assert order_again_params == expected_order_params
+
+
+def test_linear_getitem(logistic_regression_binary_pipeline_class):
+    pipeline = logistic_regression_binary_pipeline_class({'One Hot Encoder': {'top_n': 4}})
+
+    assert pipeline[0] == Imputer()
```

Comment:
Amazing! Just to confirm, giving an invalid index throws... an IndexError? Or ValueError? Just curious since get_component throws a ValueError :p

Reply:
Great question, it throws an IndexError!

![image](https://user-images.githubusercontent.com/41651716/105537458-f2fb5a80-5cbf-11eb-88a4-bc324ea205f3.png)